### PR TITLE
Add documentation for global-may-alias analysis.

### PR DESCRIPTION
**Pull request by NlightNFotis**
Document global-may-alias pointer analysis,
and add some information in the analyses/readme
file about the kind of analysis it performs.

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [x] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
-  Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- [x] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

File: src/analyses/global_may_alias.cpp [hunk 1]
```diff
@@ -11,6 +11,11 @@ Author: Daniel Kroening, kroening@kroening.com
 
 #include "global_may_alias.h"
 
+/// Populate list of aliases for a given identifier in
+/// an LHS expression.
+/// \param lhs: A left hand side expression, containing an identifier.
+/// \param alias_set: A external set of aliases to populate internal
```
Comment (xbauch) [inline src/analyses/global_may_alias.cpp]: An external set

File: src/analyses/global_may_alias.h [hunk 2]
```diff
@@ -88,6 +99,9 @@ class global_may_alias_domaint:public ai_domain_baset
   void get_rhs_aliases_address_of(const exprt &, std::set<irep_idt> &);
 };
 
+/// This is a may analysis (i.e. aliasing may occur during execution, but is
```
Comment (xbauch) [inline src/analyses/global_may_alias.h]: aliasing that may occur

File: src/analyses/global_may_alias.cpp [hunk 3]
```diff
@@ -28,6 +33,11 @@ void global_may_alias_domaint::assign_lhs_aliases(
   }
 }
 
+/// Populate list of aliases for a given identifier in
+/// an RHS expression.
```
Comment (xbauch) [inline src/analyses/global_may_alias.cpp]: a RHS

**Review by xbauch — APPROVED** [review]

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 40be8bfdbfa5d1dcb7f699940dfffad7d813b677).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/91250713

File: src/analyses/global_may_alias.cpp [hunk 4]
```diff
@@ -11,6 +11,11 @@ Author: Daniel Kroening, kroening@kroening.com
 
 #include "global_may_alias.h"
 
+/// Populate list of aliases for a given identifier in
+/// an LHS expression.
```
Comment (tautschnig) [inline src/analyses/global_may_alias.cpp]: Maybe we should avoid the use of left-hand side/LHS and right-hand side and instead be explicit and take about contexts where an object is being written to vs objects are only being read?

File: src/analyses/global_may_alias.h [hunk 5]
```diff
@@ -28,6 +28,7 @@ class global_may_alias_domaint:public ai_domain_baset
   {
   }
 
+  /// AI domain transform function.
```
Comment (tautschnig) [inline src/analyses/global_may_alias.h]: Let's spell out abstract interpretation.

**Review by tautschnig — APPROVED** [review]

File: src/analyses/README.md [hunk 6]
```diff
@@ -73,7 +73,13 @@ To be documented.
 
 \subsection analyses-global-may-alias Global may-alias analysis (global_may_aliast)
 
-To be documented.
+This is a pointer alias analysis (analysing the memory locations a pointer
+expression may point to, and finding when two pointer expressions refer to 
+the same storage locations). It is flow-insensitive, meaning that it computes
+what memory locations pointer expressions may refer to at any time during the program
+execution. It's called may-alias, because as it looks for aliasing that may occur
```
Comment (peterschrammel) [inline src/analyses/README.md]: because as -> because

File: src/analyses/README.md [hunk 7]
```diff
@@ -73,7 +73,13 @@ To be documented.
...
+during program execution (compared to analysis for aliasing that must occur). It's
```
Comment (peterschrammel) [inline src/analyses/README.md:81]: You can say that it is 'over-approximating'

File: src/analyses/README.md [hunk 8]
```diff
@@ -73,7 +73,13 @@ To be documented.
...
+performing this analysis for global values.
```
Comment (peterschrammel) [inline src/analyses/README.md]: What are global values?

File: src/analyses/global_may_alias.cpp [hunk 9]
```diff
@@ -28,6 +33,11 @@ void global_may_alias_domaint::assign_lhs_aliases(
   }
 }
 
+/// Populate list of aliases for a given identifier in
+/// an RHS expression.
+/// \param rhs: A right hand side expression.
+/// \param alias_set: A external set of aliases to populate internal
+/// aliases set.
```
Comment (peterschrammel) [inline src/analyses/global_may_alias.cpp]: I'd indent this to make it more readable (several occurrences)

File: src/analyses/global_may_alias.h (hunk 5 above)
Comment (peterschrammel) [inline src/analyses/global_may_alias.h]: You could use \copydoc to avoid repeating the documentation from ai.h

**Comment by NlightNFotis** [comment]
@peterschrammel Can you re-review this, see if everything's okay? Should be good to go after this, (assuming CI passes).

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: fc932dab352f0ab8be2a92342a01b47877826481).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/91546755

**Review by peterschrammel — APPROVED** [review]

**Pull request merged**